### DISCUSSION TITLE
Is it normal that I should restart Rails to see component preview changes?

### DISCUSSION THREAD

**Discussion by Jackson200** [lookbook-hq/lookbook#148] [Q&A]
Hi all,

Any changes I make to my `_preview.rb` files in `/spec/components` require me to restart rails to see my changes on Lookbook.

eg. If I add 
```
#
# @param group_id select [38, 42, 43]
#
```
I will need to restart Rails instead of just refreshing the page to see the params select dropdown.

I'm not convinced that it is a Lookbook problem, since I have a similar (sometimes) issue with my vanilla viewComponent previews.

I have looked for spring running, a cache that shouldn't be on, etc, but come up blank. Any ideas?

Thanks in advance.

**Comment by allmarkedup** [comment]
Hey @Jackson200 -  that is certainly not how things should be working!

When you make a change, is it causing the server to hang? Or can you still navigate around in Lookbook ok, but just not see the changes?

Also, what version of Lookbook are you running?

**Reply by Jackson200** [reply]
Hi @allmarkedup 
ruby 3.0.1 and lookbook 1.0.4
I can navigate without hanging.
I double checked http://localhost:3000/rails/view_components/ and I am missing the same new previews, so I think i may well be the underlying previews that are not updated, so rather a viewComponent issue I'm having than LookBook

**Reply by allmarkedup** [reply]
@Jackson200 ok thanks for the info - if you are not seeing them in http://localhost:3000/rails/view_components/ then it doesn't sound like a Lookbook issue. Sometimes a syntax error in a component or preview file can cause them not to show up but unless it is that I'm not sure why that might be happening. Sorry I can't be of more help!

**Reply by Jackson200** [reply]
thanks for looking @allmarkedup 👍

**Reply by aarongough** [reply]
Hi @allmarkedup! Just found LookBook and I'm very excited to use it!

I just installed it in our app. Viewing components and clicking around seems to work fine, but previews in LookBook do not update when the preview file is changed or the component itself is changed. The changes ARE reflected in `http://localhost:3000/rails/view_components/` and I can see ActionCable firing the events that seem like they should trigger a reload in LookBook:

```bash
listen: raw changes: [[:modified, "/Users/USER/work/CLIENT/spec/components/previews/chevron_toggle_component_preview.rb"]]
listen: final changes: {:modified=>["/Users/USER/work/CLIENT/spec/components/previews/chevron_toggle_component_preview.rb"], :added=>[], :removed=>[]}
[ActionCable] Broadcasting to reload: {}
Callback took 0.01683700003195554 sec
```
However the changes do not show up in LookBook until I kill the server and restart with `bundle exec rails s`... Clicking the `refresh` button in the top right of the LookBook UI does not show the changes either, neither does refreshing the whole page.

It seems like it is caching the results of rendering the previews when the server first starts, and is then never updating them?

I am using `Rails 6.1.7` on `ruby 3.1.2p20 (2022-04-12 revision 4491bb740a) [arm64-darwin21]`

Please let me know what other info would be helpful for debugging!
-Aaron

**Reply by allmarkedup** [reply]
Hey @aarongough sorry you are running into problems and thanks for the info. Obviously something funny is going on here so would be good to get to the bottom of it.

In the top right of the UI is a little **?** icon in a circle which opens a dropdown debug menu:

<img width="211" alt="CleanShot 2022-10-08 at 21 44 16" src="https://user-images.githubusercontent.com/126726/194727253-ee961698-1e75-435a-ad5f-4e7a00bd4f0e.png">

Could you click the 'copy debug data' link and paste in the output here for me to look at? It shouldn't contain any sensitive information (apart from potentially the file paths) but obviously redact anything you feel you need to :-)

Also, just to check - I assume you are running Lookbook in the  `development` environment when you are testing this?

**Reply by aarongough** [reply]
No problem at all mate! I really appreciate that you released LookBook open source, and I know how much it is to maintain a project like this! 🙂 

Here is the debug data: https://gist.github.com/aarongough/3c8a17a56565bc1de72db56a6a1cf3ab

Please note I redacted some parts of the paths (`USER` and `CLIENT`) just to protect the identity of the client I am currently working for. Let me know if there are any problems or if you need anything else!

**Reply by aarongough** [reply]
And yes, I am seeing this issue in `development`, we are also running LookBook in `staging` but there is no issue there as the server gets re-launched there's a code change...

**Comment by allmarkedup** [comment]
@aarongough Thanks for the debug data dump and sorry once again I'm so slow in replying...

Another question - are you using a forking server like Puma? And if so is your `WEB_CONCURRENCY` value set to more than `1`?

Unfortunately Lookbook's file watching system doesn't play nicely with forked processes due to a [limitation of the underlying library](https://github.com/allmarkedup/lookbook/issues/98#issuecomment-1253796251) that is used to watch for changes. So if you are using Puma with more than one process locally you will see the behaviour that you are describing.

Setting `WEB_CONCURRENCY` to 1 will fix it if that is indeed the cause of your issues (not a great solution I realise but it's the only way to work around it that I have found!).

Let me know if that works for you anyway if/when you get a chance to look at it.

**Comment by patmisch** [comment]
I'm having this problem as well, and none of the suggested fixes seem to work. Even stranger, my lookbook nav area doesn't update, even after a server restart. In order to get the nav to reflect changes to my `preview` directory or `_preview.rb` files, I have to completely re-install lookbook. 

Perhaps useful: When I click "Copy debug data" I get the following error: `TypeError: undefined is not an object (evaluating 'window.navigator.clipboard.writeText')`

**Reply by patmisch** [reply]
Turns out adding `Rails.application.load_server` to `config.ru` solves the issue of `_preview.rb` files not reloading, but I'm still seeing the odd nav behavior where the only way to get it to update is re-installing the lookbook gem.

**Reply by allmarkedup** [reply]
@patmisch thank you for following up on this - you are correct, `Rails.application.load_server` needs to be present in the `config.ru` file, this is the default for new Rails installs but I'm guessing you have upgraded from v5.0 or previous along the way?

This needs to be documented so I'll add a note in the troubleshooting section of the docs and hopefully spare other people the same hassle. Sorry you had to spend time figuring it out!

**Reply by patmisch** [reply]
@allmarkedup Any ideas on why my nav does not update without a full re-install of lookbook?

**Reply by allmarkedup** [reply]
@patmisch have you got caching enabled in development? The sidebar is fragment cached for performance reasons with the assumption that caching will be disabled in dev.

**Reply by patmisch** [reply]
Caching is disabled... But yeah, definitely seems like that must be what is going on.

**Reply by amirali-ashraf** [reply]
@allmarkedup and @patmisch thanks for the solution.
To fix your issue, I had a kind of the same issue and this might be because of zeitwerk. Could you check the naming of your files and your classes to match properly? `button_preview.rb` => `class ButtonPreview; end`. And also note the module names. Let me know if that helps.

**Comment by cj** [comment]
@allmarkedup  I just run into this issue with the latest version of rails and lookbook, here's my debug log https://gist.github.com/cj/2ef79973652a533cadbb1f1f4e33148d. If I change the following value, I have to do a full rails restart to see the changes reflected. 
<img width="671" alt="image" src="https://user-images.githubusercontent.com/1819/216195581-e82b975b-9fca-4bdf-8820-e604bed7bc3c.png">

**Reply by cj** [reply]
I managed to get it working with the following flags `-e development --no-dev-caching`

**Comment by allmarkedup** [comment]
Hey everyone, thanks for all your input on this. I think there are actually a number of different issues affecting people here, although  caching does seem to be a likely culprit for a lot of these issues.

I've got a [PR up here](https://github.com/ViewComponent/lookbook/pull/308) that adds some very basic cache-invalidation logic in to hopefully ensure updates are still seen even if caching is enabled in development.

If anyone has a spare few minutes to give that branch a test and let me know if it helps at all that would be much appreciated!
